lookout/core/bytes_to_unicode_converter.py

```diff
@@ -0,0 +1,128 @@
+from typing import Dict
+
+import bblfsh
+from lookout.sdk.service_data_pb2 import Change, File
+import numpy
+
+from lookout.core.analyzer import UnicodeChange, UnicodeFile
+
+
+class BytesToUnicodeConverter:
+    """Utility class to convert bytes positions to unicode positions in `bblfsh.Node`."""
+
+    def __init__(self, content: bytes):
+        """
+        Initialize a new instance of BytesToUnicodeConverter.
+
+        :param content: Code byte representation.
+        """
+        self._content = content
+        self._content_str = content.decode(errors="replace")
+        self._lines = self._content_str.splitlines(keepends=True)
+        self._byte_to_str_offset = self._build_bytes_to_str_offset_mapping(content)
+        self._lines_offset = self._build_lines_offset_mapping(self._content_str)
+
+    def convert_content(self):
+        """Convert byte content (or code) to unicode."""
+        return self._content_str
+
+    def convert_uast(self, uast: bblfsh.Node) -> bblfsh.Node:
+        """
+        Convert uast Nodes bytes position to unicode position.
+
+        UAST is expected to correspond to provided content.
+        :param uast: corresponding UAST.
+        :return: UAST with unicode positions.
+        """
+        uast = bblfsh.Node.FromString(uast.SerializeToString())
+        if not self._content:
+            return uast
+        for node in self._traverse_uast(uast):
+            for position in (node.start_position, node.end_position):
+                if position.offset == 0 and position.col == 0 and position.line == 0:
+                    continue
+                new_position = self._get_position(self._byte_to_str_offset[position.offset])
+                for attr in ("offset", "line", "col"):
+                    setattr(position, attr, getattr(new_position, attr))
+        return uast
+
+    def _get_position(self, offset: int) -> bblfsh.Position:
+        """Get new position for unicode string offset."""
+        line_num = numpy.argmax(self._lines_offset > offset) - 1
+        col = offset - self._lines_offset[line_num]
+        line = self._lines[line_num]
+        if len(line) == col:
+            if line.splitlines()[0] != line:
+                # ends with newline
+                line_num += 1
+                col = 0
+        return bblfsh.Position(offset=offset, line=line_num + 1, col=col + 1)
+
+    @staticmethod
+    def _build_lines_offset_mapping(content: str) -> numpy.ndarray:
```

Comment:
How about using "columns" instead of "lines_offset" or "lines_start_offsets"

Reply:
Do you mean renaming? but is it about offsets, not columns.

Reply:
An offset inside a line is called a column, which I think you update here in the function. If this is not what the function is doing, you really need to write a good docstring and also change the names to express what is really happening.

Reply:
It calculates an offset of line beginning:

So mapping from line number to offset of the 1st symbol of this line

Reply:
ok